### ISSUE TITLE
MPI_SERIAL needs MPI_VERSION

### ISSUE THREAD

**Issue by gold2718** [ESMCI/cime#1870]
The MPI serial library (imported via MCT) needs MPI_VERSION and MPI_SUBVERSION constants to allow users to write MPI2 and MPI3 code not supported by MPI_SERIAL.
This issue is waiting on PR MCSclimate/MCT#50.

**Comment by jedwards4b** [comment]
Why do this instead of adding the needed MPI2 or MPI3 code to mpi-serial?

**Comment by gold2718** [comment]
First, these constants (and the function, MPI_Get_Version) have been MPI required constants since at least MPI 1.3.
Second, many of the missing functionality is time consuming to implement (e.g., one-sided communication from MPI2) and there is no one with time to do the implementations (as far as I know).
This change will at least allow coders to use MPI2 and MPI3 functionality in CIME-supported models.

**Comment by rljacob** [comment]
Note that mpi-serial was never intended to be a complete replacement for MPI.  The right solution is to get users to install a real MPI library on their laptop or whatever.

**Comment by jedwards4b** [comment]
In order for cime/cesm testing to work mpi-serial needs to work with cime and with cesm components.   In general I think that it would be better to add functionality to the mpi-serial library rather than adding cpp macros (ifdef blocks) to the code.

**Comment by rljacob** [comment]
Knock yourself out with adding functionality and yes we don't want to macro away the MPI parts of cime.   I'm saying eventually you're going to find it easier to just change the cime/cesm requirements to "must have an MPI library installed".

**Comment by ekluzek** [comment]
I think mpi-serial is an important functionality that I'd like to see continue to be supported. Lot's of CLM people use it for single point simulations, and it does allow the system to be put together in the simplest way possible. Adding a real MPI library adds quite a bit of complexity to not only the build, but the run step as well. If you are porting to a new system or a laptop for example, it's going to be easiest to get mpi-serial working first.

Now, you are right, eventually we are going to have to implement MPI2 features in mpi-serial or it won't be useful anymore. But, if @gold2718 has a solution that works for now I think that's awesome. Another way forward at that point would be to add stub interfaces to the MPI2 features, but add an abortrun if they are called, so we know they aren't being used. That obviously wouldn't take much time. And I would hope that even implementing an interface for a single task wouldn't be that hard to do either.

**Comment by gold2718** [comment]
@ekluzek,
Some of the missing MPI2 functions are not trivial to implement in a serial library. For instance, the one-sided communication functions require possibly complex data movement even in a serial context.

**Comment by gold2718** [comment]
Is there any resolution to this discussion? This issue is still waiting on PR MCSclimate/MCT#50.

**Comment by gold2718** [comment]
@rljacob & @mvertens:

I still have not heard of any resolution to this discussion (and PR MCSclimate/MCT#50 has now been open for 17 days). This issue is holding up CAM development since we can't bring our new code to the trunk until this fix to MPI_SERIAL is in CIME.

I can submit a PR directly into CIME to avoid forcing MCT to take code they do not want or are not ready to accept. Any objections?

**Comment by jedwards4b** [comment]
I think that if it's holding up your development you should put a "good enough" solution in CIME and we can resolve it when the mct team is ready.

**Comment by rljacob** [comment]
I just haven't had time to do my testing.  I promised to finish that PR by tomorrow.   But yes you could always push need-now MCT changes directly to CIME and then do the proper submodule update later.

**Comment by mvertens** [comment]
@gold2718 - I agree with Jim.

**Comment by gold2718** [comment]
> I promised to finish that PR by tomorrow.

Did this happen? If not, do you have a procedure defined for

> do the proper submodule update later

?
In particular, is there a better way for me to put these changes into CIME that eases any required cleanup?

**Comment by rljacob** [comment]
Done now.

I meant "proper subtree update".  You could put a time-critical change in ESMCI/CIME's MCT or mpi-serial with a normal development branch.  That shouldn't interfere with a "subtree merge" done later.

**Comment by gold2718** [comment]
Thanks but there seems to be a bug which has crept into MCT since the last merge into CIME.
It seems to be related to Pat Worley's changes. I note that GNU does not pick up this error but Intel does.

The test is:
`SMS_D_Ln9.f19_g16_rx1.A.yellowstone_intel`
although I also tried this on other machines. 
The error is:

```
  14:forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable
 RSENDBUF when it is not allocated
  14:
  14:Image              PC                Routine            Line        Source 
            
  14:cesm.exe           00000000017E39A0  Unknown               Unknown  Unknown
  14:cesm.exe           00000000012F1BA2  m_rearranger_mp_r        1165  m_Rearranger.F90
  14:cesm.exe           00000000012A9FFD  m_matattrvectmul_         514  m_MatAttrVectMul.F90
  14:cesm.exe           0000000000D5F4DE  shr_strdata_mod_m         754  shr_strdata_mod.F90
  14:cesm.exe           00000000008C4AE7  datm_comp_mod_mp_         667  datm_comp_mod.F90
  14:cesm.exe           00000000008C41AE  datm_comp_mod_mp_         578  datm_comp_mod.F90
  14:cesm.exe           00000000008B9525  atm_comp_mct_mp_a         169  atm_comp_mct.F90
  14:cesm.exe           0000000000483F4D  component_mod_mp_         229  component_mod.F90
  14:cesm.exe           000000000044B2C8  cime_comp_mod_mp_        1188  cime_comp_mod.F90
  14:cesm.exe           0000000000478E10  MAIN__                     63  cime_driver.F90
  14:cesm.exe           0000000000436F9E  Unknown               Unknown  Unknown
ime_comp_mod.F90
  14:cesm.exe           0000000000478E10  MAIN__                     63  cime_driver.F90
  14:cesm.exe           0000000000436F9E  Unknown               Unknown  Unknown
```

**Comment by gold2718** [comment]
I am going to abandon the subtree merge of MCT until this MCT bug is fixed. This issue will therefore be closed by PR #1922.

**Comment by rljacob** [comment]
Can you open a bug report somewhere for the MCT problem?  Include intel compiler version and how to reproduce it.

**Comment by gold2718** [comment]
I do not know how to reproduce it in MCT, only with the particular CIME test I mention above.

**Comment by rljacob** [comment]
I mean did you replace the current version of MCT in CIME with the latest and greatest to get this?

**Comment by gold2718** [comment]
Yes. I did a subtree add using the MCSclimate/MCT master.
Then, when I ran scripts_regression_tests.py, I ended up with the error above.
Rerunning that test standalone also produces the error -- at least with the Intel compiler.
